File: 17-Binary-Tree/python/03_binary_tree_postorder_traversal.py
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        stack = []
        current = root
        previous = None

        while current or stack:
            while current:
                stack.append(current)
                current = current.left

            current = stack[-1]

            if not current.right or current.right == previous:
                result.append(current.val)
                previous = current
                stack.pop()
                current = None
            else:
                current = current.right

        return result

File: 17-Binary-Tree/cpp/03-binary-tree-postorder-traversal.cpp
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<TreeNode*> st;
        TreeNode* current = root;
        TreeNode* previous = nullptr;

        while (current != nullptr || !st.empty()) {
            while (current != nullptr) {
                st.push(current);
                current = current->left;
            }

            current = st.top();

            if (current->right == nullptr || current->right == previous) {
                result.push_back(current->val);
                previous = current;
                st.pop();
                current = nullptr;
            } else {
                current = current->right;
            }
        }

        return result;
    }
};
